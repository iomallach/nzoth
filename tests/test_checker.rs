mod helpers;
use helpers::{SnapshotErrors, create_source_file};
use nzoth::{checker::Checker, parser::Parser};

fn test_valid_cases(glob_pattern: &str) {
    insta::glob!(glob_pattern, |path| {
        let contents = std::fs::read_to_string(path).expect("Expect path exists");
        let source = create_source_file(path, contents.as_str());

        let mut parser = Parser::new(&source);
        let program = parser.parse();
        assert_eq!(parser.errors().len(), 0);

        let mut checker = Checker::new(&source);
        checker.typecheck_program(program);
        let checked_program = checker.comp_unit;

        let output = format!("{:#?}", checked_program);
        insta::with_settings!({
            omit_expression => true,
            input_file => path,
            snapshot_suffix => path.file_name().unwrap().to_str().unwrap(),
        }, {
            insta::assert_snapshot!(output);
        });
    })
}

fn test_invalid_cases(glob_pattern: &str) {
    insta::glob!(glob_pattern, |path| {
        let contents = std::fs::read_to_string(path).expect("Expect path exists");
        let source = create_source_file(path, contents.as_str());

        let mut parser = Parser::new(&source);
        let program = parser.parse();

        let mut checker = Checker::new(&source);
        checker.typecheck_program(program);

        insta::with_settings!({
            omit_expression => true,
            input_file => path,
            snapshot_suffix => path.file_name().unwrap().to_str().unwrap(),
        }, {
            insta::assert_snapshot!(SnapshotErrors(&checker.errors));
        });
    })
}

#[test]
fn test_valid_let_declarations() {
    test_valid_cases("resources/checker_cases/valid_let_declarations.nz");
}

#[test]
fn test_invalid_let_declarations() {
    test_invalid_cases("resources/checker_cases/invalid_let_declarations.nz");
}

#[test]
fn test_invalid_expressions() {
    test_invalid_cases("resources/checker_cases/invalid_expressions.nz");
}
