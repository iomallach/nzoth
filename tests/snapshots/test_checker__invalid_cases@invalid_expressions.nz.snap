---
source: tests/test_checker.rs
input_file: tests/resources/checker_cases/invalid_expressions.nz
---
Can't apply `+` to `float` and `int` at line 1, column 2:
  3.4 + 1;
  ^^^
Can't apply `-` to `int` and `float` at line 2, column 2:
  1 - 0.489;
  ^
Can't apply `*` to `int` and `float` at line 3, column 2:
  2 * 2.334;
  ^
Can't apply `/` to `float` and `int` at line 4, column 2:
  43.42 / 43;
  ^^^^^
Can't apply `>` to `bool` and `int` at line 5, column 2:
  true > 1;
  ^^^^
Can't apply `<` to `int` and `float` at line 6, column 2:
  43 < 439.4;
  ^^
Can't apply `>=` to `float` and `bool` at line 7, column 2:
  432.4 >= false;
  ^^^^^
Can't apply `<=` to `bool` and `bool` at line 8, column 2:
  false <= true;
  ^^^^^
Can't apply `>` to `bool` and `bool` at line 9, column 2:
  true > false;
  ^^^^
Can't apply `==` to `bool` and `float` at line 10, column 2:
  true == 432.49;
  ^^^^
Can't apply `!=` to `int` and `bool` at line 11, column 2:
  934 != false;
  ^^^
